presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java

```diff
@@ -643,20 +644,29 @@ public PlanWithProperties visitStatisticsWriterNode(StatisticsWriterNode node, P
         @Override
         public PlanWithProperties visitTableFinish(TableFinishNode node, PreferredProperties preferredProperties)
         {
-            PlanWithProperties child = planChild(node, PreferredProperties.any());
+            PlanNode child = planChild(node, PreferredProperties.any()).getNode();
 
-            // if the child is already a gathering exchange, don't add another
-            if ((child.getNode() instanceof ExchangeNode) && ((ExchangeNode) child.getNode()).getType().equals(GATHER)) {
-                return rebaseAndDeriveProperties(node, child);
+            ExchangeNode gather;
+            // in case the input is a union (see PushTableWriteThroughUnion)
+            if (child instanceof ExchangeNode) {
+                ExchangeNode exchangeNode = (ExchangeNode) child;
+                gather = new ExchangeNode(
```

Comment:
this cannot use `gatheringOrderedExchange` ?

Reply:
The `gatheringOrderedExchange` doesn't support multiple inputs =\ Maybe i should refactor

Reply:
I was trying to factor it out, but then I'm creating a static factory method  that effectively has all the arguments that the constructor does.